collectors/beegfsmetaMetric.go

```diff
@@ -0,0 +1,207 @@
+	ExcludeMetrics    []string `json:"exclude_metrics,omitempty"`
+	ExcludeFilesystem []string `json:"exclude_filesystem"`
+}
+
+type BeegfsMetaCollector struct {
+	metricCollector
+	tags    map[string]string
+	matches map[string]string
+	config  BeegfsMetaCollectorConfig
+	skipFS  map[string]struct{}
+}
+
+func (m *BeegfsMetaCollector) Init(config json.RawMessage) error {
+	// Check if already initialized
+	if m.init {
+		return nil
+	}
+	// Metrics
+	var nodeMdstat_array = [39]string{
+		"sum", "ack", "close", "entInf",
+		"fndOwn", "mkdir", "create", "rddir",
+		"refrEn", "mdsInf", "rmdir", "rmLnk",
+		"mvDirIns", "mvFiIns", "open", "ren",
+		"sChDrct", "sAttr", "sDirPat", "stat",
+		"statfs", "trunc", "symlnk", "unlnk",
+		"lookLI", "statLI", "revalLI", "openLI",
+		"createLI", "hardlnk", "flckAp", "flckEn",
+		"flckRg", "dirparent", "listXA", "getXA",
+		"rmXA", "setXA", "mirror"}
+
+	m.name = "BeegfsMetaCollector"
+	m.setup()
+	// Set default beegfs-ctl binary
+
+	m.config.Beegfs = "/usr/bin/beegfs-ctl"
+
+	// Read JSON configuration
+	if len(config) > 0 {
+		err := json.Unmarshal(config, &m.config)
+		if err != nil {
+			return err
+		}
+	}
+	println(m.config.Beegfs)
+	//create map with possible variables
+	m.matches = make(map[string]string)
+	for _, value := range nodeMdstat_array {
+		_, skip := stringArrayContains(m.config.ExcludeMetrics, value)
+		if skip {
+			m.matches["other"] = "0"
+		} else {
+			m.matches[value] = "0"
+		}
+	}
+
+	m.meta = map[string]string{
+		"source": m.name,
+		"group":  "BeegfsMeta",
+	}
+	m.tags = map[string]string{
+		"type":       "node",
+		"filesystem": "",
+	}
+	m.skipFS = make(map[string]struct{})
+	for _, fs := range m.config.ExcludeFilesystem {
+		m.skipFS[fs] = struct{}{}
+	}
+
+	// Beegfs file system statistics can only be queried by user root
+	user, err := user.Current()
+	if err != nil {
+		return fmt.Errorf("BeegfsMetaCollector.Init(): Failed to get current user: %v", err)
+	}
+	if user.Uid != "0" {
+		return fmt.Errorf("BeegfsMetaCollector.Init(): BeeGFS file system statistics can only be queried by user root")
+	}
+
+	// Check if beegfs-ctl is in executable search path
+	_, err = exec.LookPath(m.config.Beegfs)
+	if err != nil {
+		return fmt.Errorf("BeegfsMetaCollector.Init(): Failed to find beegfs-ctl binary '%s': %v", m.config.Beegfs, err)
+	}
+	m.init = true
+	return nil
+}
+
+func (m *BeegfsMetaCollector) Read(interval time.Duration, output chan lp.CCMetric) {
+	if !m.init {
+		return
+	}
+	//get mounpoint
+	buffer, _ := ioutil.ReadFile(string("/proc/mounts"))
+	mounts := strings.Split(string(buffer), "\n")
+	var mountpoint string
+	for _, line := range mounts {
+		if len(line) == 0 {
+			continue
+		}
+		f := strings.Fields(line)
+		if strings.Contains(f[0], "beegfs_ondemand") {
+			// Skip excluded filesystems
+			if _, skip := m.skipFS[mountpoint]; skip {
+				continue
+			}
+			mountpoint = f[1]
+		}
+	}
+	m.tags["filesystem"] = mountpoint
```

Comment:
This part looks like you restrict the collector to a single BeeGFS file system. I understand that you want to evaluate `/proc/mounts` in every read to respect the `ondemand` part but why take the last match in the list of `beegfs_ondemand` files systems? Please read all not-excluded `beegfs_ondemand` file systems. You don't need to extend the global `m.tags` but can create the metric with the basic set and add `y.AddTag("filesystem", mountpoint)` before sending it to the channel. Keep the filesystem-specific counts in a separate structure (like `fsstats[mountpoint][metric]`) if you need them in multiple reads.

Reply:
I checked the BeeGFS page and it mentions multiple mounts, so it is generally possible to have more mount points.